File: chatClient.js
const fetch = require("node-fetch");
const { v4: uuidv4 } = require("uuid");
const FormData = require("form-data");
const fs = require("fs");
const dotenv = require("dotenv");
dotenv.config();

const API_KEY = process.env.API_KEY;
const BASE_URL = "https://api.on-demand.io/chat/v1";
const MEDIA_BASE_URL = "https://api.on-demand.io/media/v1";

let EXTERNAL_USER_ID = "<your_external_user_id>";
const RESPONSE_MODE = "sync";
const STOP_SEQUENCES = [];
const TEMPERATURE = 0.7;
const TOP_P = 1;
const MAX_TOKENS = 0;
const PRESENCE_PENALTY = 0;
const FREQUENCY_PENALTY = 0;

// File settings
const FILE_PATH = "<path_to_your_file>";
const FILE_NAME = "<file_name>";
const CREATED_BY = "AIREV";
const UPDATED_BY = "AIREV";
const FILE_AGENTS = [
  "agent-1713954536",
  "agent-1713958591",
  "agent-1713958830",
  "agent-1713961903",
  "agent-1713967141",
];

async function uploadMediaFile(filePath, fileName, agents, sessionId) {
  const url = `${MEDIA_BASE_URL}/public/file/raw`;

  if (!fs.existsSync(filePath)) {
    console.log(`❌ File not found: ${filePath}`);
    return null;
  }

  console.log(`\n--- Uploading Media File ---`);
  console.log(`📁 File: ${filePath}`);
  console.log(`📝 Name: ${fileName}`);
  console.log(`🔌 Agents: ${JSON.stringify(agents)}`);

  try {
    const formData = new FormData();
    formData.append("file", fs.createReadStream(filePath));
    formData.append("sessionId", sessionId);
    formData.append("createdBy", CREATED_BY);
    formData.append("updatedBy", UPDATED_BY);
    formData.append("name", fileName);
    formData.append("responseMode", RESPONSE_MODE);
    agents.forEach((agent) => {
      formData.append("agents", agent);
    });

    const response = await fetch(url, {
      method: "POST",
      headers: { apikey: API_KEY, ...formData.getHeaders() },
      body: formData,
    });

    if (response.status === 201 || response.status === 200) {
      const mediaResponse = await response.json();
      console.log(`✅ Media file uploaded successfully!`);
      console.log(`📄 File ID: ${mediaResponse.data.id}`);
      console.log(`🔗 URL: ${mediaResponse.data.url}`);

      if (mediaResponse.data.context) {
        console.log(
          `📋 Context: ${mediaResponse.data.context.substring(0, 200)}...`
        );
      }

      return mediaResponse.data;
    } else {
      const respBody = await response.text();
      console.log(
        `❌ Error uploading media file: ${response.status} - ${respBody}`
      );
      return null;
    }
  } catch (error) {
    console.log(`❌ Exception during file upload: ${error.message}`);
    return null;
  }
}

async function createChatSession(contextMetadata, config) {
  const url = `${BASE_URL}/sessions`;
  const body = {
    agentIds: config.agentIds, // Use config
    externalUserId: EXTERNAL_USER_ID,
    contextMetadata: contextMetadata,
  };

  const response = await fetch(url, {
    method: "POST",
    headers: { apikey: API_KEY, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (response.status === 201) {
    const sessionRespData = await response.json();
    return sessionRespData.data.id;
  } else {
    const txt = await response.text();
    console.log(`❌ Error creating session: ${txt}`);
    return "";
  }
}

async function submitQuery(sessionId, contextMetadata, userQuery, config) {
  const url = `${BASE_URL}/sessions/${sessionId}/query`;
  const body = {
    endpointId: config.endpointId,
    query: userQuery,
    agentIds: config.agentIds,
    responseMode: RESPONSE_MODE,
    reasoningMode: config.reasoningMode,
    modelConfigs: {
      fulfillmentPrompt: config.fulfillmentPrompt,
      stopSequences: STOP_SEQUENCES,
      temperature: TEMPERATURE,
      topP: TOP_P,
      maxTokens: MAX_TOKENS,
      presencePenalty: PRESENCE_PENALTY,
      frequencyPenalty: FREQUENCY_PENALTY,
    },
  };

  console.log(`🚀 Submitting query...`);

  const response = await fetch(url, {
    method: "POST",
    headers: { apikey: API_KEY, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (RESPONSE_MODE === "sync") {
    if (response.status === 200) {
      const original = await response.json();
      if (original.data) original.data.contextMetadata = contextMetadata;
      return original;
    }
    throw new Error(await response.text());
  } else if (RESPONSE_MODE === "stream") {
    console.log("✅ Streaming Response...");
    if (!response.body) return null;

    let fullAnswer = "";
    let finalSessionId = "";
    let finalMessageId = "";
    let metrics = {};

    // Note: ensure your node-fetch version/environment supports getReader, otherwise use .on('data')
    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    async function read() {
      const { done, value } = await reader.read();
      if (done) return;

      const chunk = decoder.decode(value, { stream: true });
      const lines = chunk.split("\n");

      for (const line of lines) {
        if (line.startsWith("data:")) {
          const dataStr = line.slice(5).trim();
          if (dataStr === "[DONE]") return;
          try {
            const event = JSON.parse(dataStr);
            if (event.eventType === "fulfillment") {
              if (event.answer) fullAnswer += event.answer;
              if (event.sessionId) finalSessionId = event.sessionId;
              if (event.messageId) finalMessageId = event.messageId;
            } else if (event.eventType === "metricsLog") {
              if (event.publicMetrics) metrics = event.publicMetrics;
            }
          } catch (e) {
            continue;
          }
        }
      }
      await read();
    }

    await read();

    const finalResponse = {
      message: "Chat query submitted successfully",
      data: {
        sessionId: finalSessionId,
        messageId: finalMessageId,
        answer: fullAnswer,
        metrics: metrics,
        status: "completed",
        contextMetadata: contextMetadata,
      },
    };

    return finalResponse;
  }
}

async function executeChat(userQuery, config) {
  // Validate config
  if (!config || !config.agentIds || !config.endpointId) {
    throw new Error("❌ Missing configuration (agentIds, endpointId, etc)");
  }
  if (API_KEY === "<your_api_key>" || !API_KEY)
    throw new Error("❌ Set API_KEY");

  if (EXTERNAL_USER_ID === "<your_external_user_id>" || !EXTERNAL_USER_ID) {
    EXTERNAL_USER_ID = uuidv4();
  }

  const contextMetadata = [
    { key: "userId", value: "1" },
    { key: "name", value: "John" },
  ];

  const sessionId = await createChatSession(contextMetadata, config);

  if (sessionId) {
    if (
      FILE_PATH !== "<path_to_your_file>" &&
      FILE_PATH &&
      fs.existsSync(FILE_PATH)
    ) {
      await uploadMediaFile(FILE_PATH, FILE_NAME, FILE_AGENTS, sessionId);
    }

    // Pass config to query submission
    const result = await submitQuery(
      sessionId,
      contextMetadata,
      userQuery,
      config
    );
    return result;
  }
  return null;
}

module.exports = executeChat;

File: Python/chat_client.py
import json
import os
import uuid
import httpx
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

API_KEY = os.getenv("API_KEY")
BASE_URL = "https://api.on-demand.io/chat/v1"
MEDIA_BASE_URL = "https://api.on-demand.io/media/v1"

# Default configuration
EXTERNAL_USER_ID = "<your_external_user_id>"
RESPONSE_MODE = "sync"  # Can be 'sync' or 'stream'
STOP_SEQUENCES = []
TEMPERATURE = 0.7
TOP_P = 1
MAX_TOKENS = 0
PRESENCE_PENALTY = 0
FREQUENCY_PENALTY = 0

# File settings
FILE_PATH = "<path_to_your_file>"
FILE_NAME = "<file_name>"
CREATED_BY = "AIREV"
UPDATED_BY = "AIREV"
FILE_AGENTS = [
    "agent-1713954536",
    "agent-1713958591",
    "agent-1713958830",
    "agent-1713961903",
    "agent-1713967141",
]


async def upload_media_file(
    client: httpx.AsyncClient, file_path, file_name, agents, session_id
):
    """
    Uploads a media file asynchronously.
    """
    url = f"{MEDIA_BASE_URL}/public/file/raw"

    if not os.path.exists(file_path):
        print(f"❌ File not found: {file_path}")
        return None

    print(f"\n--- Uploading Media File ---")
    print(f"📁 File: {file_path}")
    print(f"🔌 Agents: {agents}")

    try:
        # Prepare form data
        # httpx handles data slightly differently than requests for multipart
        data = {
            "sessionId": session_id,
            "createdBy": CREATED_BY,
            "updatedBy": UPDATED_BY,
            "name": file_name,
            "responseMode": RESPONSE_MODE,
        }

        # Add agents (array handling)
        # We manually construct the list of tuples for multi-value keys
        data_payload = list(data.items())
        for agent in agents:
            data_payload.append(("agents", agent))

        headers = {"apikey": API_KEY}

        # Open file in binary mode
        with open(file_path, "rb") as f:
            files = {"file": (os.path.basename(file_path), f)}

            response = await client.post(
                url, headers=headers, files=files, data=data_payload
            )

        if response.status_code in [200, 201]:
            media_response = response.json()
            data = media_response.get("data", {})
            print(f"✅ Media file uploaded successfully! ID: {data.get('id')}")
            return data
        else:
            print(
                f"❌ Error uploading media file: {response.status_code} - {response.text}"
            )
            return None

    except Exception as e:
        print(f"❌ Exception during file upload: {str(e)}")
        return None


async def create_chat_session(client: httpx.AsyncClient, context_metadata, config):
    """
    Creates a new chat session asynchronously.
    """
    url = f"{BASE_URL}/sessions"
    agent_ids = config.get("toolIds", [])

    body = {
        "agentIds": agent_ids,
        "externalUserId": EXTERNAL_USER_ID,
        "contextMetadata": context_metadata,
    }

    headers = {"apikey": API_KEY, "Content-Type": "application/json"}

    try:
        response = await client.post(url, json=body, headers=headers)

        if response.status_code == 201:
            session_data = response.json()
            return session_data["data"]["id"]
        else:
            print(f"❌ Error creating session: {response.text}")
            return ""
    except Exception as e:
        print(f"❌ Exception creating session: {e}")
        return ""


async def submit_query(
    client: httpx.AsyncClient, session_id, context_metadata, user_query, config
):
    """
    Submits the query asynchronously. Handles both sync and streaming.
    """
    url = f"{BASE_URL}/sessions/{session_id}/query"

    body = {
        "endpointId": config.get("endpointId"),
        "query": user_query,
        "agentIds": config.get("toolIds", []),
        "responseMode": RESPONSE_MODE,
        "reasoningMode": config.get("reasoningMode"),
        "modelConfigs": {
            "fulfillmentPrompt": config.get("fulfillmentPrompt", ""),
            "stopSequences": STOP_SEQUENCES,
            "temperature": TEMPERATURE,
            "topP": TOP_P,
            "maxTokens": MAX_TOKENS,
            "presencePenalty": PRESENCE_PENALTY,
            "frequencyPenalty": FREQUENCY_PENALTY,
        },
    }

    print(f"🚀 Submitting query...")

    headers = {"apikey": API_KEY, "Content-Type": "application/json"}

    # SYNC MODE
    if RESPONSE_MODE == "sync":
        response = await client.post(url, json=body, headers=headers)
        if response.status_code == 200:
            original = response.json()
            if "data" in original:
                original["data"]["contextMetadata"] = context_metadata
            return original
        else:
            raise Exception(f"Error from API: {response.text}")

    # STREAM MODE
    elif RESPONSE_MODE == "stream":
        print("✅ Streaming Response...")

        full_answer = ""
        final_session_id = ""
        final_message_id = ""
        metrics = {}

        # Use client.stream for streaming responses
        async with client.stream("POST", url, json=body, headers=headers) as response:
            async for line in response.aiter_lines():
                if line:
                    if line.startswith("data:"):
                        data_str = line[5:].strip()  # Remove 'data:'

                        if data_str == "[DONE]":
                            break

                        try:
                            event = json.loads(data_str)
                            if event.get("eventType") == "fulfillment":
                                if "answer" in event:
                                    full_answer += event["answer"]
                                if "sessionId" in event:
                                    final_session_id = event["sessionId"]
                                if "messageId" in event:
                                    final_message_id = event["messageId"]
                            elif event.get("eventType") == "metricsLog":
                                if "publicMetrics" in event:
                                    metrics = event["publicMetrics"]
                        except json.JSONDecodeError:
                            continue

        final_response = {
            "message": "Chat query submitted successfully",
            "data": {
                "sessionId": final_session_id,
                "messageId": final_message_id,
                "answer": full_answer,
                "metrics": metrics,
                "status": "completed",
                "contextMetadata": context_metadata,
            },
        }
        return final_response


# Main exported function
async def execute_chat(user_query, config):
    """
    The main entry point called by the specific agents.
    Uses a single httpx.AsyncClient for efficiency.
    """

    # Validation
    if not config or "toolIds" not in config or "endpointId" not in config:
        raise ValueError("❌ Missing configuration (toolIds, endpointId, etc)")

    if not API_KEY or API_KEY == "<your_api_key>":
        raise ValueError("❌ Set API_KEY in environment variables")

    global EXTERNAL_USER_ID
    if not EXTERNAL_USER_ID or EXTERNAL_USER_ID == "<your_external_user_id>":
        EXTERNAL_USER_ID = str(uuid.uuid4())

    context_metadata = [
        {"key": "userId", "value": "1"},
        {"key": "name", "value": "John"},
    ]

    # Create one client for the lifecycle of this request
    async with httpx.AsyncClient(timeout=60.0) as client:

        # 1. Create Session
        session_id = await create_chat_session(client, context_metadata, config)

        if session_id:
            # 2. Upload File (if path exists and differs from placeholder)
            if (
                FILE_PATH
                and FILE_PATH != "<path_to_your_file>"
                and os.path.exists(FILE_PATH)
            ):

                await upload_media_file(
                    client, FILE_PATH, FILE_NAME, FILE_AGENTS, session_id
                )

            # 3. Submit Query
            result = await submit_query(
                client, session_id, context_metadata, user_query, config
            )
            return result

    return None
